src/bvals/boundary_conditions_generic.hpp

```diff
@@ -34,6 +37,35 @@ namespace BoundaryFunction {
 enum class BCSide { Inner, Outer };
 enum class BCType { Outflow, Reflect, ConstantDeriv, Fixed, FixedFace };
 
+namespace impl {
+using desc_key_t = std::tuple<bool, TopologicalType>;
+template <class... var_ts>
+using map_bc_pack_descriptor_t =
+    std::unordered_map<desc_key_t, typename SparsePack<var_ts...>::Descriptor,
+                       tuple_hash<desc_key_t>>;
+
+template <class... var_ts>
+map_bc_pack_descriptor_t<var_ts...>
+GetPackDescriptorMap(std::shared_ptr<MeshBlockData<Real>> &rc) {
```

Comment:
Yeah, the way the boundary conditions work we call `GenericBC` for each type of topological element. This just builds a map from (topological element, coarse) to the corresponding pack descriptor so that the map can be `static` in `GenericBC` and pack descriptors only get built once. Rebuilding the pack descriptors was a significant performance hit for MG since the boundary stuff gets called a lot.

Reply:
:+1: